### PR TITLE
Fix download agent search relying on throwing of Sys.which(). Update t #28157

### PR DESCRIPTION
**Pull request by wfrgra**
Closes #28157.

### CODE REVIEW

File: base/download.jl [hunk 1]
```diff
@@ -25,28 +25,16 @@ if Sys.iswindows()
...
-                    # Sys.which() will throw() if it can't find `checkcmd`
-                    Sys.which(checkcmd)
-                    downloadcmd = checkcmd
-                    break
-                catch
-                end
-            end
-        end
-        if downloadcmd == "wget"
+        if !isempty(Sys.which("wget"))
```
Comment (stevengj) [inline base/download.jl]: shouldn't this be `Sys.which("wget") !== nothing`?

Comment (wfrgra) [reply base/download.jl]: Yes, not sure where I got the idea that empty strings were being returned

**Review by stevengj — APPROVED** [review]

**Comment by staticfloat** [comment]
Note that this changes behavior slightly, in that it will search the PATH every single time we invoke it, rather than caching the result as was done previously.  I personally don't care that much, as I think that's kind of a premature optimization, but it was there.

**Comment by StefanKarpinski** [comment]
Seems like a bug fix to me where the bug is "`Sys.which` does not pick up changes to `PATH`".

**Comment by stevengj** [comment]
Causing a CI failure:
```
Error in testset file:
Test Failed at /tmp/julia/share/julia/test/file.jl:878
  Expression: download("ba\0d", "good")
    Expected: ArgumentError
      Thrown: Base.IOError
ERROR: LoadError: Test run finished with errors
in expression starting at /tmp/julia/share/julia/test/runtests.jl:61
```

**Comment by KristofferC** [comment]
Can this be tested?

File: base/download.jl [hunk 2]
```diff
@@ -29,7 +29,7 @@ else
             try
                 run(`wget -O $filename $url`)
             catch
-                rm(filename)  # wget always creates a file
+                isfile(filename) && rm(filename)  # wget always creates a file
```
Comment (stevengj) [inline base/download.jl]: just do `rm(filename, force=true)`

**Pull request merged**